server/sonar-server/src/main/java/org/sonar/server/batch/ProjectDataLoader.java

```diff
@@ -0,0 +1,252 @@
+      return module;
+    }
+  }
+
+  private Map<String, String> getSettingsFromParents(ComponentDto module, boolean hasScanPerm, DbSession session) {
+    List<ComponentDto> parents = newArrayList();
+    aggregateParentModules(module, parents, session);
+    Collections.reverse(parents);
+
+    Map<String, String> parentProperties = newHashMap();
+    for (ComponentDto parent : parents) {
+      parentProperties.putAll(getPropertiesMap(dbClient.propertiesDao().selectProjectProperties(session, parent.key()), hasScanPerm));
+    }
+    return parentProperties;
+  }
+
+  private void aggregateParentModules(ComponentDto component, List<ComponentDto> parents, DbSession session) {
+    String moduleUuid = component.moduleUuid();
+    if (moduleUuid != null) {
+      ComponentDto parent = dbClient.componentDao().selectOrFailByUuid(session, moduleUuid);
+      if (parent != null) {
+        parents.add(parent);
+        aggregateParentModules(parent, parents, session);
+      }
+    }
+  }
+
+  private void addSettingsToChildrenModules(ProjectRepositories ref, String moduleKey, Map<String, String> parentProperties, TreeModuleSettings treeModuleSettings,
+    boolean hasScanPerm) {
+    Map<String, String> currentParentProperties = newHashMap();
+    currentParentProperties.putAll(parentProperties);
+    currentParentProperties.putAll(getPropertiesMap(treeModuleSettings.findModuleSettings(moduleKey), hasScanPerm));
+    addSettings(ref, moduleKey, currentParentProperties);
+
+    for (ComponentDto childModule : treeModuleSettings.findChildrenModule(moduleKey)) {
+      addSettings(ref, childModule.getKey(), currentParentProperties);
+      addSettingsToChildrenModules(ref, childModule.getKey(), currentParentProperties, treeModuleSettings, hasScanPerm);
+    }
+  }
+
+  private static void addSettings(ProjectRepositories ref, String module, Map<String, String> properties) {
+    if (!properties.isEmpty()) {
+      ref.addSettings(module, properties);
+    }
+  }
+
+  private static Map<String, String> getPropertiesMap(List<PropertyDto> propertyDtos, boolean hasScanPerm) {
+    Map<String, String> properties = newHashMap();
+    for (PropertyDto propertyDto : propertyDtos) {
+      String key = propertyDto.getKey();
+      String value = propertyDto.getValue();
+      if (isPropertyAllowed(key, hasScanPerm)) {
+        properties.put(key, value);
+      }
+    }
+    return properties;
+  }
+
+  private static boolean isPropertyAllowed(String key, boolean hasScanPerm) {
+    return !key.contains(".secured") || hasScanPerm;
+  }
+
+  private static void addFileData(ProjectRepositories data, List<ComponentDto> moduleChildren, List<FilePathWithHashDto> files) {
+    Map<String, String> moduleKeysByUuid = newHashMap();
+    for (ComponentDto module : moduleChildren) {
+      moduleKeysByUuid.put(module.uuid(), module.key());
+    }
+
+    for (FilePathWithHashDto file : files) {
+      FileData fileData = new FileData(file.getSrcHash(), file.getRevision());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) org.sonar.server.batch.ProjectDataLoader must not use org.sonar.batch.protocol.input.FileData [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AThe_Server_stack_should_not_access_classes_of_the_Batch_stack)